commit 0b2c7f89721326b4cdbf7c38607a014b8de9f9c8
Author: Albert Lee <trisk@omniti.com>
Date:   Thu Sep 3 08:54:57 2015 -0400

    Initial zones support via datasetns

diff --git a/include/sys/zfs_context.h b/include/sys/zfs_context.h
index 3dc54f1..3c2510d 100644
--- a/include/sys/zfs_context.h
+++ b/include/sys/zfs_context.h
@@ -654,9 +654,6 @@ typedef struct callb_cpr {
 	mutex_exit((cp)->cc_lockp);				\
 }
 
-#define	zone_dataset_visible(x, y)	(1)
-#define	INGLOBALZONE(z)			(1)
-
 extern char *kmem_vasprintf(const char *fmt, va_list adx);
 extern char *kmem_asprintf(const char *fmt, ...);
 #define	strfree(str) kmem_free((str), strlen(str) + 1)
